File: clients/android/NewsBlur/src/com/newsblur/activity/FeedItemsList.java
package com.newsblur.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import android.view.Menu;
import android.view.MenuItem;

import com.google.android.play.core.review.ReviewInfo;
import com.google.android.play.core.review.ReviewManager;
import com.google.android.play.core.review.ReviewManagerFactory;
import com.google.android.play.core.tasks.Task;
import com.newsblur.R;
import com.newsblur.di.IconLoader;
import com.newsblur.domain.Feed;
import com.newsblur.fragment.DeleteFeedFragment;
import com.newsblur.fragment.FeedIntelTrainerFragment;
import com.newsblur.fragment.RenameDialogFragment;
import com.newsblur.util.FeedExt;
import com.newsblur.util.FeedSet;
import com.newsblur.util.ImageLoader;
import com.newsblur.util.PrefsUtils;
import com.newsblur.util.Session;
import com.newsblur.util.SessionDataSource;
import com.newsblur.util.UIUtils;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class FeedItemsList extends ItemsList {

    @Inject
    @IconLoader
    ImageLoader iconLoader;

    public static final String EXTRA_FEED = "feed";
    public static final String EXTRA_FOLDER_NAME = "folderName";
	private Feed feed;
	private String folderName;
	private ReviewManager reviewManager;
	private ReviewInfo reviewInfo;

    public static void startActivity(Context context, FeedSet feedSet,
                                     Feed feed, String folderName,
                                     @Nullable SessionDataSource sessionDataSource) {
        Intent intent = new Intent(context, FeedItemsList.class);
        intent.putExtra(FeedItemsList.EXTRA_FEED, feed);
        intent.putExtra(FeedItemsList.EXTRA_FOLDER_NAME, folderName);
        intent.putExtra(ItemsList.EXTRA_FEED_SET, feedSet);
        intent.putExtra(ItemsList.EXTRA_SESSION_DATA, sessionDataSource);
        context.startActivity(intent);
    }

	@Override
	protected void onCreate(Bundle bundle) {
		super.onCreate(bundle);
        setupFeedItems(getIntent());
        viewModel.getNextSession().observe(this, this::setupFeedItems);
        checkInAppReview();
    }

    @Override
    public void onBackPressed() {
        // see checkInAppReview()
        if (reviewInfo != null) {
            Task<Void> flow = reviewManager.launchReviewFlow(this, reviewInfo);
            flow.addOnCompleteListener(task -> {
                PrefsUtils.setInAppReviewed(this);
                super.onBackPressed();
            });
        } else {
            super.onBackPressed();
        }
    }

    public void deleteFeed() {
		DialogFragment deleteFeedFragment = DeleteFeedFragment.newInstance(feed, folderName);
		deleteFeedFragment.show(getSupportFragmentManager(), "dialog");
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (super.onOptionsItemSelected(item)) {
            return true;
        }
        if (item.getItemId() == R.id.menu_delete_feed) {
            deleteFeed();
            return true;
        }
        if (item.getItemId() == R.id.menu_notifications_disable) {
            feedUtils.disableNotifications(this, feed);
            return true;
        }
        if (item.getItemId() == R.id.menu_notifications_focus) {
            feedUtils.enableFocusNotifications(this, feed);
            return true;
        }
        if (item.getItemId() == R.id.menu_notifications_unread) {
            feedUtils.enableUnreadNotifications(this, feed);
            return true;
        }
        if (item.getItemId() == R.id.menu_instafetch_feed) {
            feedUtils.instaFetchFeed(this, feed.feedId);
            this.finish();
            return true;
        }
        if (item.getItemId() == R.id.menu_intel) {
            FeedIntelTrainerFragment intelFrag = FeedIntelTrainerFragment.newInstance(feed, fs);
            intelFrag.show(getSupportFragmentManager(), FeedIntelTrainerFragment.class.getName());
            return true;
        }
        if (item.getItemId() == R.id.menu_rename_feed) {
            RenameDialogFragment frag = RenameDialogFragment.newInstance(feed);
            frag.show(getSupportFragmentManager(), RenameDialogFragment.class.getName());
            return true;
            // TODO: since this activity uses a feed object passed as an extra and doesn't query the DB,
            // the name change won't be reflected until the activity finishes.
        }
        if (item.getItemId() == R.id.menu_statistics) {
            feedUtils.openStatistics(this, feed.feedId);
            return true;
        }
        return false;
	}

	@Override
	public boolean onPrepareOptionsMenu(Menu menu) {
		super.onPrepareOptionsMenu(menu);
        if (FeedExt.isAndroidNotifyUnread(feed)) {
            menu.findItem(R.id.menu_notifications_disable).setChecked(false);
            menu.findItem(R.id.menu_notifications_unread).setChecked(true);
            menu.findItem(R.id.menu_notifications_focus).setChecked(false);
        } else if (FeedExt.isAndroidNotifyFocus(feed)) {
            menu.findItem(R.id.menu_notifications_disable).setChecked(false);
            menu.findItem(R.id.menu_notifications_unread).setChecked(false);
            menu.findItem(R.id.menu_notifications_focus).setChecked(true);
        } else {
            menu.findItem(R.id.menu_notifications_disable).setChecked(true);
            menu.findItem(R.id.menu_notifications_unread).setChecked(false);
            menu.findItem(R.id.menu_notifications_focus).setChecked(false);
        }
		return true;
	}

    @Override
    String getSaveSearchFeedId() {
        return "feed:" + feed.feedId;
    }

    private void setupFeedItems(Session session) {
        Feed feed = session.getFeed();
        String folderName = session.getFolderName();
        if (feed != null && folderName != null) {
            setupFeedItems(feed, folderName);
        } else {
            finish();
        }
    }

    private void setupFeedItems(Intent intent) {
        Feed feed = (Feed) intent.getSerializableExtra(EXTRA_FEED);
        String folderName = intent.getStringExtra(EXTRA_FOLDER_NAME);
        setupFeedItems(feed, folderName);
    }

    private void setupFeedItems(@NonNull Feed feed, @NonNull String folderName) {
        this.feed = feed;
        this.folderName = folderName;
        UIUtils.setupToolbar(this, feed.faviconUrl, feed.title, iconLoader, false);
    }

    private void checkInAppReview() {
        if (!PrefsUtils.hasInAppReviewed(this)) {
            reviewManager = ReviewManagerFactory.create(this);
            Task<ReviewInfo> request = reviewManager.requestReviewFlow();
            request.addOnCompleteListener(task -> {
                if (task.isSuccessful()) {
                    reviewInfo = task.getResult();
                }
            });
        }
    }
}

File: clients/android/NewsBlur/app/src/main/java/com/newsblur/activity/FeedItemsList.kt
package com.newsblur.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.fragment.app.DialogFragment
import com.newsblur.R
import com.newsblur.database.BlurDatabaseHelper
import com.newsblur.di.IconLoader
import com.newsblur.domain.Feed
import com.newsblur.fragment.AddFeedFragment
import com.newsblur.fragment.DeleteFeedFragment
import com.newsblur.fragment.FeedIntelTrainerFragment
import com.newsblur.fragment.RenameDialogFragment
import com.newsblur.service.NbSyncManager.UPDATE_METADATA
import com.newsblur.util.CustomIconRenderer
import com.newsblur.util.FeedExt.isAndroidNotifyFocus
import com.newsblur.util.FeedExt.isAndroidNotifyUnread
import com.newsblur.util.FeedSet
import com.newsblur.util.ImageLoader
import com.newsblur.util.Session
import com.newsblur.util.SessionDataSource
import com.newsblur.util.UIUtils
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class FeedItemsList : ItemsList() {
    @Inject
    @IconLoader
    lateinit var iconLoader: ImageLoader

    private lateinit var feed: Feed
    private lateinit var folderName: String
    private var isTryFeed = false
    private var tryFeedUrl: String? = null

    private val reviewHelper by lazy { InAppReviewHelper(this, prefsRepo) }

    override fun onCreate(bundle: Bundle?) {
        super.onCreate(bundle)
        setupFeedItems(intent)

        viewModel.nextSession.observe(this) { session ->
            setupFeedItems(session)
        }

        updateTryFeedBanner()
        reviewHelper.check()
    }

    override fun interceptBackPress(isGestureNavigation: Boolean): Boolean =
        reviewHelper.interceptBackPress(isGestureNavigation) { finish() }

    override fun shouldResetReadingSessionOnCreate(): Boolean =
        intent?.getBooleanExtra(EXTRA_IS_TRY_FEED, false) == true

    override fun handleUpdate(updateType: Int) {
        super.handleUpdate(updateType)
        if ((updateType and UPDATE_METADATA) != 0) {
            refreshFeedHeader()
        }
    }

    fun showDeleteFeedDialog() {
        val deleteFeedFragment: DialogFragment = DeleteFeedFragment.newInstance(feed, folderName)
        deleteFeedFragment.show(supportFragmentManager, "dialog")
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (super.onOptionsItemSelected(item)) {
            return true
        }

        return when (item.itemId) {
            R.id.menu_delete_feed -> {
                showDeleteFeedDialog()
                true
            }

            R.id.menu_notifications_disable -> {
                feedUtils.disableNotifications(this, feed)
                true
            }

            R.id.menu_notifications_focus -> {
                feedUtils.enableFocusNotifications(this, feed)
                true
            }

            R.id.menu_notifications_unread -> {
                feedUtils.enableUnreadNotifications(this, feed)
                true
            }

            R.id.menu_instafetch_feed -> {
                feedUtils.instaFetchFeed(this, feed.feedId)
                finish()
                true
            }

            R.id.menu_intel -> {
                val intelFrag = FeedIntelTrainerFragment.newInstance(feed, fs)
                intelFrag.show(supportFragmentManager, FeedIntelTrainerFragment::class.java.name)
                true
            }

            R.id.menu_rename_feed -> {
                val frag = RenameDialogFragment.newFeedInstance(feed.feedId, feed.title)
                frag.show(supportFragmentManager, RenameDialogFragment::class.java.name)
                true
            }

            R.id.menu_statistics -> {
                feedUtils.openStatistics(this, prefsRepo, feed.feedId)
                true
            }

            else -> false
        }
    }

    override fun prepareItemListMenuModel(menu: Menu): Boolean {
        super.prepareItemListMenuModel(menu)
        if (!::feed.isInitialized) return true

        when {
            feed.isAndroidNotifyUnread() -> {
                menu.findItem(R.id.menu_notifications_disable).isChecked = false
                menu.findItem(R.id.menu_notifications_unread).isChecked = true
                menu.findItem(R.id.menu_notifications_focus).isChecked = false
            }

            feed.isAndroidNotifyFocus() -> {
                menu.findItem(R.id.menu_notifications_disable).isChecked = false
                menu.findItem(R.id.menu_notifications_unread).isChecked = false
                menu.findItem(R.id.menu_notifications_focus).isChecked = true
            }

            else -> {
                menu.findItem(R.id.menu_notifications_disable).isChecked = true
                menu.findItem(R.id.menu_notifications_unread).isChecked = false
                menu.findItem(R.id.menu_notifications_focus).isChecked = false
            }
        }
        return true
    }

    override fun getSaveSearchFeedId(): String = if (::feed.isInitialized) "feed:${feed.feedId}" else ""

    private fun setupFeedItems(session: Session) {
        val feed = session.feed
        val folderName = session.folderName
        if (feed != null && folderName != null) {
            setupFeedItems(feed, folderName)
        } else {
            finish()
        }
    }

    private fun setupFeedItems(intent: Intent) {
        val feed = intent.getSerializableExtra(EXTRA_FEED) as Feed?
        val folderName = intent.getStringExtra(EXTRA_FOLDER_NAME)
        isTryFeed = intent.getBooleanExtra(EXTRA_IS_TRY_FEED, false)
        tryFeedUrl = intent.getStringExtra(EXTRA_TRY_FEED_URL)
        if (feed != null && folderName != null) {
            setupFeedItems(feed, folderName)
        } else {
            finish()
        }
    }

    private fun setupFeedItems(
        feed: Feed,
        folderName: String,
    ) {
        this.feed = feed
        this.folderName = folderName
        val customIcon = BlurDatabaseHelper.getFeedIcon(feed.feedId)
        if (customIcon != null) {
            val iconSize = UIUtils.dp2px(this, 24)
            val iconBitmap = CustomIconRenderer.renderIcon(this, customIcon, iconSize)
            if (iconBitmap != null) {
                UIUtils.setupToolbar(this, iconBitmap, feed.title, false)
                return
            }
        }
        UIUtils.setupToolbar(this, feed.faviconUrl, feed.title, iconLoader, false)
        updateTryFeedBanner()
    }

    private fun refreshFeedHeader() {
        if (!::feed.isInitialized || !::folderName.isInitialized) return
        dbHelper.getFeed(feed.feedId)?.let { updatedFeed ->
            setupFeedItems(updatedFeed, folderName)
        }
        updateTryFeedBanner()
    }

    private fun updateTryFeedBanner() {
        if (!isTryFeed || !::feed.isInitialized) {
            binding.itemlistTryFeedBanner.visibility = android.view.View.GONE
            return
        }

        val palette = com.newsblur.util.discoverThemePalette(this, prefsRepo)
        binding.itemlistTryFeedBanner.visibility = android.view.View.VISIBLE
        val background = android.graphics.drawable.GradientDrawable()
        background.shape = android.graphics.drawable.GradientDrawable.RECTANGLE
        background.cornerRadius = 0f
        background.setColor(palette.tryFeedBannerBackgroundColor)
        background.setStroke(UIUtils.dp2px(this, 1), palette.tryFeedBannerBorderColor)
        binding.itemlistTryFeedBanner.background = background
        binding.itemlistTryFeedTitle.text = feed.title
        binding.itemlistTryFeedSubtitle.setText(R.string.try_feed_banner_subtitle)
        binding.itemlistTryFeedTitle.setTextColor(palette.tryFeedBannerTitleColor)
        binding.itemlistTryFeedSubtitle.setTextColor(palette.tryFeedBannerSubtitleColor)
        binding.itemlistTryFeedSubscribeButton.backgroundTintList = android.content.res.ColorStateList.valueOf(palette.tryFeedButtonBackgroundColor)
        binding.itemlistTryFeedSubscribeButton.setTextColor(palette.tryFeedButtonTextColor)
        binding.itemlistTryFeedSubscribeButton.setOnClickListener {
            val feedUrl = tryFeedUrl?.takeIf { it.isNotBlank() } ?: feed.address.takeIf { it.isNotBlank() } ?: feed.feedLink
            if (!feedUrl.isNullOrBlank()) {
                AddFeedFragment
                    .newInstance(feedUrl, feed.title, clearTryFeedOnSuccess = true)
                    .show(supportFragmentManager, AddFeedFragment::class.java.name)
            }
        }
        iconLoader.displayImage(feed.faviconUrl, binding.itemlistTryFeedIcon)
    }

    companion object {
        const val EXTRA_FEED: String = "feed"
        const val EXTRA_FOLDER_NAME: String = "folderName"
        const val EXTRA_IS_TRY_FEED: String = "is_try_feed"
        const val EXTRA_TRY_FEED_URL: String = "try_feed_url"

        @JvmStatic
        fun startActivity(
            context: Context,
            feedSet: FeedSet,
            feed: Feed?,
            folderName: String?,
            sessionDataSource: SessionDataSource?,
        ) {
            Intent(context, FeedItemsList::class.java)
                .apply {
                    putExtra(EXTRA_FEED, feed)
                    putExtra(EXTRA_FOLDER_NAME, folderName)
                    putExtra(EXTRA_FEED_SET, feedSet)
                    putExtra(EXTRA_SESSION_DATA, sessionDataSource)
                }.also { intent ->
                    context.startActivity(intent)
                }
        }

        @JvmStatic
        fun startStoryActivity(
            context: Context,
            feedSet: FeedSet,
            feed: Feed?,
            folderName: String?,
            storyHash: String,
        ) {
            Intent(context, FeedItemsList::class.java)
                .apply {
                    putExtra(EXTRA_FEED, feed)
                    putExtra(EXTRA_FOLDER_NAME, folderName)
                    putExtra(EXTRA_FEED_SET, feedSet)
                    putExtra(EXTRA_STORY_HASH, storyHash)
                    putExtra(EXTRA_AUTO_OPEN_STORY, true)
                }.also { intent ->
                    context.startActivity(intent)
                }
        }

        @JvmStatic
        fun startTryFeedActivity(
            context: Context,
            feed: Feed,
        ) {
            Intent(context, FeedItemsList::class.java)
                .apply {
                    putExtra(EXTRA_FEED, feed)
                    putExtra(EXTRA_FOLDER_NAME, com.newsblur.util.AppConstants.ROOT_FOLDER)
                    putExtra(EXTRA_FEED_SET, FeedSet.singleFeed(feed.feedId))
                    putExtra(EXTRA_IS_TRY_FEED, true)
                    putExtra(EXTRA_TRY_FEED_URL, feed.address)
                }.also { intent ->
                    context.startActivity(intent)
                }
        }
    }
}
